torch/csrc/jit/tensorexpr/analysis.h

```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
   bool found_{false};
 };
 
+// This indicates the path from root to a stmt A in AST. Specifically, it
+// consists of a vector of integers: each integer represents the number of stmts
+// before A's ancestor or A in the block. For example, "1:1:0" points to stmt
+// "d[i] += e[i, j];" in the following code.
+//
+// c = a/b;
+// for i
+//  d[i] = 0;
+//  for j
+//    d[i] += e[i, j];
+class StmtTrace {
+ public:
+  void append(int32_t id) {
+    trace_.emplace_back(id);
+  }
+  void pop() {
+    trace_.pop_back();
+  }
+
+  const std::vector<int32_t>& getTrace() {
+    return trace_;
+  }
+
+  bool operator==(StmtTrace& trace) {
+    auto compare = trace.getTrace();
+    if (trace_.size() != compare.size()) {
+      return false;
+    }
+    int size = compare.size();
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) != compare.at(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool operator<(StmtTrace& trace) {
+    auto compare = trace.getTrace();
+    int size = trace_.size() < compare.size() ? trace_.size() : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) > compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == trace);
+  }
+
+  bool operator>(StmtTrace& trace) {
+    auto compare = trace.getTrace();
+    int size = trace_.size() < compare.size() ? trace_.size() : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) < compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == trace);
+  }
+
+  std::string getTraceString() {
+    std::string cstr = "";
+    for (int i = 0; i < trace_.size(); i++) {
+      cstr += std::to_string(trace_.at(i));
+      cstr += ":";
+    }
+    return cstr;
+  }
+
+ private:
+  std::vector<int32_t> trace_;
+};
+
+// Traverse the root stmt to identify the position of each stmt in the AST,
+// i.e., a StmtTrace
+class StmtTracer : public IRVisitor {
+ public:
+  StmtTrace getStmtTrace() {
+    return trace_;
+  }
+
+ private:
+  void visit(BlockPtr v) {
+    if (flatten_) {
+      for (StmtPtr s : *v) {
+        s->accept(this);
+      }
+      return;
+    }
+
+    int count = 0;
+    for (StmtPtr s : *v) {
+      trace_.append(count);
+      s->accept(this);
+      trace_.pop();
+      count++;
+    }
+  }
+
+  void visit(CondPtr v) {
+    flatten_ = true;
+    ExprPtr condition = v->condition();
+    StmtPtr true_stmt = v->true_stmt();
+    StmtPtr false_stmt = v->false_stmt();
+    condition->accept(this);
+    if (true_stmt) {
+      true_stmt->accept(this);
+    }
+    if (false_stmt) {
+      false_stmt->accept(this);
+    }
+    flatten_ = false;
+  }
+
+  StmtTrace trace_;
+  bool flatten_ = false;
+};
+
+enum AccMode { READ, WRITE, REDUCE };
+
+// Traverses the IR to identify all reads/writes to a buf, and their positions
+// in the AST which is represented as a StmtTrace
+using BufAccessInfo = std::tuple<StmtPtr, AccMode, StmtTrace>;
```

Comment:
StmtTrace can be omitted from this tuple since we can always obtain it when we have StmtPtr. Generally it's better to pass as little (yet enough) information as possible to keep the API simpler.

Reply:
Also, we already have two `AccessInfo` classes:
* https://github.com/pytorch/pytorch/blob/efdb17b984dc0624290d3a943d18df78ba5f55ed/torch/csrc/jit/tensorexpr/registerizer.h#L51
* https://github.com/pytorch/pytorch/blob/efdb17b984dc0624290d3a943d18df78ba5f55ed/torch/csrc/jit/tensorexpr/mem_dependency_checker.h#L38

I think we really need to first try using them, or merging them, or deleting some of them before adding a third one.

Reply:
>StmtTrace can be omitted from this tuple since we can always obtain it when we have StmtPtr.

I don't think so. Each `StmtPtr` is a node in the AST of the root stmt. We don't have their `StmtIndex` somewhere but have to visit the root stmt again to identify its position. Here, we have identified the position of all buf access stmts in one visit and are passing this info. I think it's better to re-compute their positions by visiting the root stmt multiple times, one time for each access.